File: un6unw/Week24/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] wires) {
        int minDifference = Integer.MAX_VALUE;

        // 각 전선을 하나씩 끊어보면서 두 개의 네트워크 크기 차이를 계산
        for (int i = 0; i < wires.length; i++) {
            // 그래프 생성 (인접 리스트)
            List<List<Integer>> graph = new ArrayList<>();
            for (int j = 0; j <= n; j++) {
                graph.add(new ArrayList<>());
            }

            // 간선 추가 (단, i번째 간선은 제외)
            for (int j = 0; j < wires.length; j++) {
                if (j == i) continue; // 현재 끊을 간선은 추가하지 않음
                int a = wires[j][0];
                int b = wires[j][1];
                graph.get(a).add(b);
                graph.get(b).add(a);
            }

            // BFS 또는 DFS를 이용해 한쪽 네트워크 크기를 구함
            boolean[] visited = new boolean[n + 1];
            int count = getNetworkSize(graph, visited, 1); // 임의의 노드(1)에서 시작

            // 다른 네트워크 크기는 n - count
            int difference = Math.abs((n - count) - count);
            minDifference = Math.min(minDifference, difference);
        }

        return minDifference;
    }

    private int getNetworkSize(List<List<Integer>> graph, boolean[] visited, int node) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(node);
        visited[node] = true;
        int count = 1;

        while (!queue.isEmpty()) {
            int current = queue.poll();
            for (int next : graph.get(current)) {
                if (!visited[next]) {
                    visited[next] = true;
                    queue.add(next);
                    count++;
                }
            }
        }
        return count;
    }
}

File: oyatplum/Week24/전력망을 둘로 나누기.py
count = 0
def dfs(v, graph, visited, n):
    global count
    count += 1
    visited[v] = 1
    for i in range(n+1):
        if graph[v][i] == 1 and visited[i] == 0:
            dfs(i, graph, visited, n)
    return count

def solution(n, wires):
    graph = [[0]*(n+1) for _ in range(n+1)] #그래프 초기화
    visited = [] #방문한 노드
    temp = [] #절대값 저장 배열
    global count #노드 개수

    for i in wires: #전력망 끊기 전 모든 간선 추가
        graph[i[0]][i[1]] = 1
        graph[i[1]][i[0]] = 1

    for i in wires: #간선마다 돌아가며 제거
        graph[i[0]][i[1]] = 0 #여기가 제거하는 로직 0으로
        graph[i[1]][i[0]] = 0

        visited = [0] * (n+1) #방문한 노드 초기화
        a = dfs(i[0], graph, visited, n) #i[0]간선 첫번째 노드 기준으로 트리 돌기
        count = 0 #노드 개수 초기화

        visited = [0] * (n+1) #방문한 노드 초기화
        b = dfs(i[1], graph, visited, n) #i[1]간선 두번째 노드 기준으로 트리 돌기
        count = 0 #노드 개수 초기화

        temp.append(abs(a-b))

        graph[i[0]][i[1]] = 1 #제거한 간선 다시 추가!
        graph[i[1]][i[0]] = 1

    return min(temp)